pulsar-client-cpp/lib/ProducerImpl.cc

```diff
@@ -550,10 +559,14 @@ void ProducerImpl::batchMessageTimeoutHandler(const boost::system::error_code& e
         return;
     }
     LOG_DEBUG(getName() << " - Batch Message Timer expired");
-    Lock lock(mutex_);
-    auto failures = batchMessageAndSend();
-    lock.unlock();
-    failures.complete();
+
+    // ignore if the producer is already closing/closed
+    if (assertState(Ready)) {
+        Lock lock(mutex_);
```

Comment:
should the assertState not happen inside the lock?

Reply:
I see that assertState take the lock, but there's a window between that returning and it being taken again where the producer could be closed.

Reply:
I've fixed that and added some extra state_ checks before batchMessageAndSend to avoid race conditions.